eradiate/radprops/vert_dist.py

```diff
@@ -0,0 +1,409 @@
+    wherein the particles fraction is evaluated.
+
+    The vertical distribution is normalised so that:
+
+    .. math::
+        \sum_i f_i = 1
+
+    where :math:`f_i` is the particles fraction in the layer division :math:`i`.
+    """
+    bottom = documented(
+        pinttr.ib(
+            units=ucc.deferred("length"),
+            converter=pinttr.converters.to_units(ucc.deferred("length")),
+            validator=pinttr.validators.has_compatible_units,
+        ),
+        doc="Layer bottom altitude.\n\nUnit-enabled field (default: ucc[length]).",
+        type="float",
+    )
+    top = documented(
+        pinttr.ib(
+            units=ucc.deferred("length"),
+            converter=pinttr.converters.to_units(ucc.deferred("length")),
+            validator=pinttr.validators.has_compatible_units,
+        ),
+        doc="Layer top altitude.\n\nUnit-enabled field (default: ucc[length]).",
+        type="float",
+    )
+
+    def __attrs_post_init__(self):
+        if self.bottom >= self.top:
+            raise ValueError("bottom altitude must be lower than top altitude")
+
+    @classmethod
+    def from_dict(cls, d: dict) -> VerticalDistribution:
+        """
+        Initialise a :class:`VerticalDistribution` from a dictionary.
+        """
+        return cls(**d)
+
+    @property
+    @abstractmethod
+    def fractions(self) -> Callable:
+        """
+        Returns a callable that evaluates the particles fractions in the
+        layer, given an array of altitude values.
+        """
+        pass
+
+    @staticmethod
+    def _normalise(x: np.ndarray) -> np.ndarray:
+        """
+        Scale the values so that their sum is 1.
+
+        Parameter ``x`` (array):
+            Values to normalise.
+
+        Returns → array:
+            Normalised array.
+        """
+        _norm = np.sum(x)
+        if _norm > 0.0:
+            return x / _norm
+        else:
+            raise ValueError(f"Cannot normalise fractions because the norm is " f"0.")
+
+
+class VerticalDistributionFactory(BaseFactory):
+    """
+    This factory constructs objects whose classes are derived from
+    :class:`VerticalDistribution`.
+
+    .. admonition:: Registered factory members
+       :class: hint
+
+       .. factorytable::
+          :factory: VerticalDistributionFactory
+    """
+
+    _constructed_type = VerticalDistribution
+    registry = {}
+
+
+@VerticalDistributionFactory.register("uniform")
+@parse_docs
+@attr.s
+class Uniform(VerticalDistribution):
+    r"""
+    Uniform vertical distribution.
+
+    The uniform probability distribution function is:
+
+    .. math::
+        f(z) = \frac{1}{z_{\rm top} - z_{\rm bottom}}, \quad
+        z \in [z_{\rm top}, z_{\rm bottom}]
+
+    where :math:`z_{\rm top}` and :math:`z_{\rm bottom}` are the layer top and bottom
+    altitudes, respectively.
+    """
+
+    @property
+    def fractions(self) -> Callable:
+        """Return a callable that evaluates the particle fractions."""
```

Comment:
I'd be curious to see this in action: how is that better than just implementing a regular method?

Reply:
I don't think it's better. I have no idea why I made it return a callable. I'll turn that into an `eval_fractions` method, that seems just better.